### PR TITLE
CheckFirmware: allow for update of public keys for secure boot

### PR DESCRIPTION
**Pull request by tridge**
This implements the SECURE_COMMAND mavlink operations. See:
 https://github.com/ArduPilot/mavlink/pull/284
MAVProxy implementation here:
 https://github.com/ArduPilot/MAVProxy/pull/1101

This allows for:
 - fetching of current public keys
 - setting new public keys
 - removing public keys

all operations are done using signed commands with an existing private key checked against the current public keys

Other changes:
 - allow boot if no public keys
 - added --signing-key to configure for easier development and upload
 - added CI test of signing
 - prevent flashing a bootloader with no public keys from secure fw

### CODE REVIEW

**Review by timtuxworth — COMMENTED** [review]
There are multiple uses of goto here. Surely this can/should be avoided?

File: libraries/AP_CheckFirmware/AP_CheckFirmware_secure_command.cpp [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+void AP_CheckFirmware::handle_secure_command(mavlink_channel_t chan, const mavlink_secure_command_t &pkt)
+{
+    mavlink_secure_command_reply_t reply {};
+    reply.result = MAV_RESULT_UNSUPPORTED;
+    reply.sequence = pkt.sequence;
+    reply.operation = pkt.operation;
+
+    if (uint16_t(pkt.data_length) + uint16_t(pkt.sig_length) > sizeof(pkt.data)) {
+        reply.result = MAV_RESULT_DENIED;
+        goto send_reply;
```
[the remark this answers is not in the document]
Comment (tridge) [reply libraries/AP_CheckFirmware/AP_CheckFirmware_secure_command.cpp:275]: goto is perfectly ok for error handling like this, it is actually a lot more readable than if()

File: libraries/AP_CheckFirmware/AP_CheckFirmware.cpp [hunk 2]
```diff
@@ -14,25 +14,39 @@
 #include <string.h>
 #include "monocypher.h"
 
-struct PACKED secure_data {
-    uint8_t sig[8] = {0x4e, 0xcf, 0x4e, 0xa5, 0xa6, 0xb6, 0xf7, 0x29};
-    struct PACKED {
-        uint8_t key[32] = {};
-    } public_key[10];
-};
+const struct ap_secure_data public_keys __attribute__((section(".ecc_raw")));
```
Comment (bugobliterator) [inline libraries/AP_CheckFirmware/AP_CheckFirmware.cpp:17]: ecc_raw name should be changed to maybe public key or something, as we are not using ECC cryptography anymore.

**Review by bugobliterator — CHANGES_REQUESTED** [review]
Looks good, just a minor name change and then good to go in.

File: libraries/AP_CheckFirmware/AP_CheckFirmware.cpp (hunk 2 above)
Comment (tridge) [reply libraries/AP_CheckFirmware/AP_CheckFirmware.cpp:17]: ok, will do

Comment (tridge) [reply libraries/AP_CheckFirmware/AP_CheckFirmware.cpp]: fixed

**Review by bugobliterator — APPROVED** [review]

**Pull request merged**